Introduce additional types of exceptions next to `unhandled` exceptions (#10)
getsentry/rfcs: text/0010-exception-types.md

- `handled = true` and `process_terminated = false`: Exception was reported via `Sentry.capture*()` method. Same as `handled` in the list above.

In the absence of the `handled` or its value being null, it's assumed to be `handled = true`. This is also the current behavior.

The introduction of the `process_terminated` flag enables the consideration of such exception types in the `session health` metric.

# Option 2

This one is very similar to option 1, however instead of an additional flag, this introduces an enum for the different types.
Once again, the mechanism needs to be adapted:

```json
{
  "exception": {
    "values": [
      {
        "type": "Error",
        "value": "An error occurred",
        "mechanism": {
          "type": "generic",
          "exception_type": "handled|unhandled|process_termination", // <--- newly introduced field
        }
      }
    ]
  }
}
```

If the currently available `handled` flag is also present, the `exception_type` flag takes precedence. The `handled` flag however should become deprecated.

> The introduction of the `exception_type` flag enables the consideration of such exception types in the `session health` metric.

Reviewer: The release health protocol would need to change as well, because right now we only have `crashed` and `errored`
https://docs.sentry.io/product/releases/health/#session-status
but we need to represent 3 states.

Reviewer 2: It seems to me we don't really need to change Sessions. We can just go from errroed to crashed in case process_exit=true instead of handled:false

# Option 3

Unhandled exceptions, which don't cause a process termination, are considered like exceptions which cause the process to terminate and are marked `handled: false`
That would however make it impossible to differentiate between those exception types in the `session health` metric.

# Approaches taken by other monitoring tools

- Crashlytics just differentiates between manually caught exceptions and unhandled exceptions, regardless of wether they cause the process to terminate.

# List of SDK to which this applies

This list might be incomplete

- [Flutter](https://github.com/getsentry/sentry-dart/issues/456)
- Browser SDKs
- Unity

> - React Natice

Reviewer: ```suggestion
- React Native
```

* Start Date: 2022-09-17
* RFC Type: feature
* RFC PR: 0010
* RFC Status: draft

# Summary

This RFC suggests a feature which introduces additional types of exceptions next to `mechanism.handled`.

# Motivation

Currently, exception which cause the running software to exit (the process died/hard crash) are marked as `handled: false`. This isn't enough for SDKs where an exception can be unhandled but at the same time doesn't cause the software to exit.

> Sentry shows exceptions which aren't caught by the developer (unhandled) but also do not cause the software to exit in the same way as exceptions which are manually caught by the developer. This seems rather unintuitive and makes exceptions seem less severe than they are.

Reviewer 2: this seem inaccurate to me. Or maybe I'm not understanding.

Sentry doesn't show exceptions that are not captured by the user (unhandled) in the same way it shows exceptions  exceptions manually captured (handled). One has the red banner:
<img width="108" alt="image" src="https://user-images.githubusercontent.com/1633368/192897344-49dcdbfe-d67c-49cd-ace3-0b11b40652c8.png">

Do you mean that Sentry shows all unhandled errors in the same way, independent if they caused the app to crash?

Author: Unhandled exceptions in Flutter don't have that red banner because they don't crash the application. I'm not sure how similar behavior is implemented in other SDKs.

That means there's no visual differentiation between handled and unhandled exceptions as long as they don't cause the software to exit.

> Do you mean that Sentry shows all unhandled errors in the same way, independent if they caused the app to crash?

No, unhandled exceptions which cause the software to exit are shown visually different.

Reviewer: Right now, Handled and Unhandled exceptions are `handled: null` or `handled: true` so there's no red banner, no way to filter or alert them differently.
Only when it crashes on the Native side (C code or Native threads), the flag is `handled: false`.
So yes, it's bad and the RFC tries to fix this.

This issue impacts for example the [Dart/Flutter SDK](https://github.com/getsentry/sentry-dart/issues/456), the Unity SDK, the React Native SDK and possibly more.

> Another issue is, that excpetions which don't cause the software to exit but are unhandled, are not considered in the [`session health` metric](https://develop.sentry.dev/sdk/sessions/).

Reviewer 2: What do you mean by 'they are not considered'?

My understanding is that we mark a session as crashed for all unhandled errors. For example, on Browser, we finish the session as crashed if there's an unhandled error, even though there was no crash.

Is this referring to Flutter where in the event of a Dart unhandled exception we don't close the session as crashed? But if there's a native unhandled error (which terminates the app) we do?

Author: > My understanding is that we mark a session as crashed for all unhandled errors. For example, on Browser, we finish the session as crashed if there's an unhandled error, even though there was no crash.

That probably wouldn't be the case for Flutter, if it had sessions, based on the current behavior. Even for Flutter for Web.

> Is this referring to Flutter where in the event of a Dart unhandled exception we don't close the session as crashed? But if there's a native unhandled error (which terminates the app) we do?

Yes, exactly.

Based on your questions, I'm assuming that `crashed`/`unhandled` mean different things across different SDKs. That's one more reason to align across SDKs.

Reviewer: > Based on your questions, I'm assuming that crashed/unhandled mean different things across different SDKs. That's one more reason to align across SDKs.

Yes, this RFC tries to align this inconsistency as well @ueman .

Sessions are only marked as crashes when it cashes on the Native side.

Currently, the session would be marked as `errored` instead of `crashed`.

> The attribute `thread.errored` was added in the past for similar reasons, but it got [reverted](https://github.com/getsentry/relay/pull/306) for some reasons.

Reviewer 2: ```suggestion
The attribute `thread.errored` was added in the past for similar reasons, but it got [reverted on Relay](https://github.com/getsentry/relay/pull/306) and [Android](https://github.com/getsentry/sentry-android/pull/187).
```

- `handled = true` and `process_terminated = true`: Software was gracefully shut down after an handled exception. This should never happen and is invalid.

In the absence of the `handled` or its value being null, it's assumed to be `handled = true`. This is also the current behavior.

The introduction of the `process_terminated` flag enables the consideration of such exception types in the `session health` metric.

The [session protocol](https://docs.sentry.io/product/releases/health/#session-status) would need to change as well though, because there are only `errored` and `crashed` states.

# Option 2

This one is very similar to option 1, however instead of an additional flag, this introduces an enum for the different types.
Once again, the mechanism needs to be adapted:

```json
{
  "exception": {
    "values": [
      {
        "type": "Error",
        "value": "An error occurred",
        "mechanism": {
          "type": "generic",
          "exception_type": "handled|unhandled|process_termination", // <--- newly introduced field
        }
      }
    ]
  }
}
```

> If the currently available `handled` flag is also present, the `exception_type` flag takes precedence. The `handled` flag however should become deprecated.

Reviewer 2: This flag is how users current filter and set alerts. Deprecating this feel like a big change

The introduction of the `exception_type` flag enables the consideration of such exception types in the `session health` metric.

# Option 3

Unhandled exceptions, which don't cause a process termination, are considered like exceptions which cause the process to terminate and are marked `handled: false`
That would however make it impossible to differentiate between those exception types in the `session health` metric, filters, alerts, etc.

# Approaches taken by other monitoring tools

> - Crashlytics just differentiates between manually caught exceptions and unhandled exceptions, regardless of wether they cause the process to terminate.

Reviewer 2: ```suggestion
- Crashlytics just differentiates between manually caught exceptions and unhandled exceptions, regardless of whether they cause the process to terminate.
```

# List of SDK to which this applies

This list might be incomplete

- [Flutter](https://github.com/getsentry/sentry-dart/issues/456)
- Browser SDKs
- Unity

> - React Native

Reviewer 2: ```suggestion
- React Native
- .NET (UnobservedTaskException)
```

- `process termination`: The exception was recorded automatically by Sentry through the use of a exception handler or similar. The exception caused the software to terminate the execution. This should be visualized in the Sentry user interface. This is currently done by the `handled: false` flag in the [exception mechanism](https://develop.sentry.dev/sdk/event-payloads/exception/#exception-mechanism).

A user of Sentry should be able to 

- filter events on the issues page or discover for the newly introduces exception types (3 categories).
- highlight (similar to the unhandled label) events of the type unhandled and process termination.
- get alerted for events of the type unhandled and process termination separately.

Currently, there's an unhandled label on the issue's page but it's only highlighted for process termination errors (if `handled: false`).

> In order to propagate those exception types, the exception mechanism needs to be adapted:

Reviewer 3: I like this option more, because it is backwards compatible. 
But I would not add another boolean, the new "process_terminated" variable needs to be added because "handled" is a boolean and thus not very future proof. 
So I would add a number/enum kind of type called something more general like `error_state` or such that can then have values like `process_terminated`, `timeout`, `deadlock`, `caught-fire`, ...

Reviewer: > One general question: How can we send data to sentry, when the process terminates? Is this even possible? (can we store the error fast enough to disk to send on next start?)

yes

          "type": "generic",
          "exception_type": "handled|unhandled|process_termination", // <--- newly introduced field
        }
      }
    ]
  }
}
```

If the currently available `handled` flag is also present, the `exception_type` flag takes precedence. The `handled` flag however should become deprecated.

The introduction of the `exception_type` flag enables the consideration of such exception types in the `session health` metric.

# Option 3

Unhandled exceptions, which don't cause a process termination, are considered like exceptions which cause the process to terminate and are marked `handled: false`
That would however make it impossible to differentiate between those exception types in the `session health` metric, filters, alerts, etc.

# Approaches taken by other monitoring tools

- Crashlytics just differentiates between manually caught exceptions and unhandled exceptions, regardless of whether they cause the process to terminate.

# List of SDK to which this applies

This list might be incomplete

- [Flutter](https://github.com/getsentry/sentry-dart/issues/456)
- Browser SDKs
- Unity
- React Native
- .NET (UnobservedTaskException)

# Unresolved Questions

> - Are there any other exception types next to the ones metioned in this RFC?

Reviewer: ```suggestion
- Are there any other exception types next to the ones metioned in this RFC?
- When there's an unhandled exception, we'd automatically finish the transaction bound to the scope, but only if the app crashes (process termination)? See [issue](https://github.com/getsentry/develop/issues/443).
```

* Start Date: 2022-09-17
* RFC Type: feature
* RFC PR: 0010
* RFC Status: draft

# Summary

> This RFC suggests a feature which introduces additional types of exceptions next to `unhandled`.

Reviewer: ```suggestion
This RFC suggests a feature which introduces additional types of exceptions next to `mechanism.handled`.
```

Flag `unhandled` does not exist so it's confusing, I've updated the description and issue's title as well.

# Motivation

> Currently, exception which cause the running software to exit are marked as `unhandled: true`. This isn't enough for SDKs where an exception can be unhandled but at the same time doesn't cause the software to exit.

Reviewer: ```suggestion
Currently, exception which cause the running software to exit are marked as `handled: false`. This isn't enough for SDKs where an exception can be unhandled but at the same time doesn't cause the software to exit.
```

> Sentry shows exceptions which aren't caught but also do not cause the software to exit in the same way as exceptions which are manually caught by a developer. This seems rather unintuitive and makes exceptions seem less severe than they are.

Reviewer: ```suggestion
Sentry shows exceptions which aren't caught by the developer but also do not cause the software to exit in the same way as exceptions which are manually caught by the developer. This seems rather unintuitive and makes exceptions seem less severe than they are.
```

This issue impacts for example the [Dart/Flutter SDK](https://github.com/getsentry/sentry-dart/issues/456), the Unity SDK, the React Native SDK and possibly more.

> Another issue is, that excpetions which don't cause the application to exit but are uncaught, are not considered in the [`session health` metric](https://develop.sentry.dev/sdk/sessions/).

Reviewer: ```suggestion
Another issue is, that excpetions which don't cause the application to exit but are unhandled, are not considered in the [`session health` metric](https://develop.sentry.dev/sdk/sessions/).
Currently, the session would be marked as `errored` instead of `crashed`.
```

> # Proposal

Reviewer: I'd add 2 more items to the proposal:

A way to filter events on the issues page or discover, such as:

`handled:yes`, `handled:no` are already possible.
Introducing a new one which filters the `process termination` type, eg `handled:termination` or a brand new tag, that's just an example.

The other 2 is the UI label highlight for `unhandled` events and `process termination`.
Currently, there's an `unhandled` label on the issue's page but it's only highlighted for `process termination` errors.

Author: Added ✅

> Based on the problem stated above, I propose to introduce the types of `handled`, `unhandled`, `process termination` (this is the same as the current `unhandeld`, but rephrased to avoid confusion). I'm open for better phrasing of those types, but I'll stick to those names for the rest of the RFC. The meaning of those types is as follows:

Reviewer: ```suggestion
Based on the problem stated above, I propose to introduce the types of `handled`, `unhandled`, `process termination` (this is the same as the current `handled`, but rephrased to avoid confusion). I'm open for better phrasing of those types, but I'll stick to those names for the rest of the RFC. The meaning of those types is as follows:
```

- `handled`: The exception was recorded by a developer via `Sentry.capture*` method. May or may not be visually indicated by the Sentry user interface.
- `unhandled`: Indicates whether the exception was recorded automatically by Sentry through the use of a global exception handler or similar. This exception however didn't cause the software to exit, and the software will continue to be executed. This should be visualized in the Sentry user interface.

> - `process termination`: The exception was recorded automatically by Sentry through the use of a exception handler or similar. The exception caused the software to terminate the execution. This should be visualized in the Sentry user interface. This is currently done by the `unhandled` flag in the [exception mechanism](https://develop.sentry.dev/sdk/event-payloads/exception/#exception-mechanism).

Reviewer: ```suggestion
- `process termination`: The exception was recorded automatically by Sentry through the use of a exception handler or similar. The exception caused the software to terminate the execution. This should be visualized in the Sentry user interface. This is currently done by the `handled` flag in the [exception mechanism](https://develop.sentry.dev/sdk/event-payloads/exception/#exception-mechanism).
```

In order to propagate those exception types, the exception mechanism needs to be adapted:

```json
{
  "exception": {
    "values": [
      {
        "type": "Error",
        "value": "An error occurred",
        "mechanism": {
          "type": "generic",
          "handled": true,
          "process_terminated": false // <--- newly introduced field
        }
      }
    ]
  }
}
```

In order to achieve backwards compatibility, in the absence of the `process_termination` flag, the current behavior stays as is.
As soon as the `process_terminated` flag is present the bavior is as follows:

> - `handled = true` and `process_terminated = true`: Software was gracefully shut down after an handled exception

Reviewer: I believe this will never happen, if it's handled, the application won't terminate.

Author: I just added this to cover all 4 possible cases. I guess if it happens, it's the developer setting it manually.

Reviewer: Gotcha, but it's an invalid use case, I'd rename it as "invalid" or even remove it, otherwise it's just more information to digest that does not represent a real use case.

> - `handled = true` and `process_terminated = false`: Exception is not handled by the user but didn't cause the software to terminate. Same as `unhandled` in the list above

Reviewer: ```suggestion
- `handled = false` and `process_terminated = false`: Exception is not handled by the user but didn't cause the software to terminate. Same as `unhandled` in the list above
```

- `handled = false` and `process_terminated = true`: Software terminated after an unhandled exception. Same as `process termination` in the list above
- `handled = true` and `process_terminated = false`: Exception was reported via `Sentry.capture*()` method. Same as `handled` in the list above.

> In the absence of the `handled` or its value being null, it's assumed to be `handled = true`.

Reviewer: That's already the current behavior.

Author: I added a note for that.

The introduction of the `process_terminated` flag enables the consideration of such exception types in the `session health` metric.

> I'm guessing this affects data ingestion layer, but since I'm not familiar with that part, I can't comment on the impact theses changes would have on that.

Reviewer: I don't think so, because clients (SDKs) are responsible for sending the `status` of the sessions, the SDKs check the type of the exception and mark them as such.
The question is if we want to introduce a new category for release health, which differentiates `handled`, `unhandled`, and `application exist`.
Right now there are only 2 categories, `errored` and `crashed`.
`crashed` meaning `process termination`.
`errored` meaning `handled` and `unhandled`.

Author: Removed this sentence

> # Other options considered

Reviewer: Usually, an RFC consists of different options, one of them being the recommended one.

So I'd just rename them to `Option #1 (recommended), `Option #2`, etc...

Author: Done ✅

> # Other options considered

Reviewer: I'd add an option which is to deprecate the `handled` flag and introduce a new one which is an `enum` with the 3 known types.
We'd send the `handled` flag still for a while.
Ingestion, SDKs and UI would use this new flag in case it exists otherwise it uses the current `handled` flag because of back compatibility.

Reviewer: I'd add another option which is setting `handled: false` for unhandled errors that don't cause the application to crash.
Sessions would be marked as `crashed`.
No changes would be required on the ingestion and UI, but obviously, you'd not be able to differentiate such events.
Maybe that would be better than now, but I agree it's not the best case.

Author: Added ✅

Unhandled exceptions, which don't cause a process termination, are considered like exceptions which cause the process to terminate.

> That would however make it impossible to differentiate between those exception types in the `session health` metric.

Reviewer: Some SDKs do that, but they are languages/platforms that never cause the application to exist, such as the Browser.
Unhandled JS events are marked as `handled: false` so sessions are `crashed`, but the browser never exit.

# Approaches taken by other monitoring tools

> - Crashlytics just differentiates between manually caught exceptions and unhandled exceptions, regardless of wether they cause the process to terminate.

Reviewer: The problem here is that you cannot give a higher priority to events that cause the application to exit.

# Unresolved Questions

> - Which SDKs would profit from this RFC? Is it the majority, a good chunk of it, or just the minory?

Reviewer: Likely every language/platform/framework that either does not exit at all (such as the Browser SDK), or when they swallow the exception such as Flutter and a few other technologies.

Author: Added a list of known SDKs which would profit from this RFC

> - Are there any other exception types next to the ones metioned in this RFC?

Reviewer: I guess this RFC covers the 3 known types.
Manually captured by the dev.
Automatically captured via error handlers, not causing the app to exit.
Automatically captured via error handlers, causing the app to exit.
